app/samples/phase_detection/src/main.cpp

```diff
@@ -0,0 +1,241 @@
+// F-Core Includes
+#include "f_core/util/debouncer.hpp"
+#include "f_core/util/linear_fit.hpp"
+
+#include <f_core/device/sensor/c_accelerometer.h>
+#include <f_core/device/sensor/c_barometer.h>
+#include <f_core/device/sensor/c_gyroscope.h>
+#include <f_core/device/sensor/c_magnetometer.h>
+#include <math.h>
+#include <zephyr/kernel.h>
+
+// Zephyr Includes
+#include "flight.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, CONFIG_APP_PHASE_DETECT_LOG_LEVEL);
+
+#include "f_core/util/debouncer.hpp"
+
+#include <array>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+
+K_TIMER_DEFINE(imu_timer, NULL, NULL);
+K_TIMER_DEFINE(barom_timer, NULL, NULL);
+
+Controller controller{sourceNames, eventNames, timer_events, deciders};
+
+void imu_thread_f(void *, void *, void *) {
+    // > 5G-ish for a quarter second
+    constexpr uint32_t boost_time_ms = 250;
+    constexpr double boost_threshold_mps2 = 50;
+
+    // < 3.5G-ish for a quarter second
+    // weird since this example is doing what grim did and doing it based on magnitude not a specific axis
+    // Deceleration is still acceleration which is why we cant just say < 1G
+    constexpr uint32_t coast_time_ms = 250;
+    constexpr double coast_threshold_mps2 = 35;
+    Debuouncer<ThresholdDirection::Over, double> boost_detector{boost_time_ms, boost_threshold_mps2};
+    Debuouncer<ThresholdDirection::Under, double> coast_detector{coast_time_ms, coast_threshold_mps2};
+
+    CAccelerometer acc(*DEVICE_DT_GET_ONE(openrocket_imu));
+    if (!acc.IsReady()) {
+        LOG_WRN("Accelerometer not ready");
+    }
+    controller.SubmitEvent(Sources::IMU1, Events::PadReady);
+    controller.WaitUntilEvent(Events::PadReady);
+
+    while (true) {
+        k_timer_status_sync(&imu_timer);
+        uint32_t timestamp = k_uptime_get();
+        bool good = acc.UpdateSensorValue();
+        if (!good) {
+            LOG_ERR("Failure reading imu");
+        }
+        double x = acc.GetSensorValueDouble(SENSOR_CHAN_ACCEL_X);
+        double y = acc.GetSensorValueDouble(SENSOR_CHAN_ACCEL_Y);
+        double z = acc.GetSensorValueDouble(SENSOR_CHAN_ACCEL_Z);
+        double mag = sqrt(x * x + y * y + z * z);
+
+        // Boost Detecting
+        if (!controller.HasEventOccured(Events::Boost)) {
```

Comment:
Depending on things, the math might be not directly run on sensor thread but rather on some zbus callback or other such decoupling mechanism. However, as of now all, phase detections algos use are O(1) and a pretty small constant at that so its not impossible to run them at the same speed that we are measuring.

For more complicated GNC, we may have to subsample and think harder on other threads but for now we're quick